README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# petro.One

[![Travis-CI Build
Status](https://travis-ci.org/f0nzie/petro.One.svg?branch=master)](https://travis-ci.org/f0nzie/petro.One)
[![codecov](https://codecov.io/gh/f0nzie/petro.One/branch/master/graph/badge.svg)](https://codecov.io/gh/f0nzie/petro.One)

The goal of **petro.One** is providing a reproducible platform for
acquiring and analyzing metadata by searching papers on oil and gas from
the OnePetro website.

## Motivation

The standard way of searching for papers in
[OnePetro](https://www.onepetro.org/) is using a web browser entering
the search terms for a particular paper we are looking for. The result
will come as web pages containing dozens, hundreds or thousand of paper
titles. We need to browse all the resulting pages to find the papers
that have a good match with the subject we are researching. Although it
has the capability to export to *csv* and *xls* files, they will need to
be imported into another application to perform some analysis.

By using some statistical tools available through `R`, the search could
turn in highly profitable in terms of time, matching quality and
selection of the papers. The search keywords are entered thorugh the R
console and the papers will return in a shape of a dataframe, which is
identical to a spreadsheet: rows of paper titles and columns with
details from the metadata extracted from the web page. With the
dataframe already in our computers we could perform a thorough search
and narrow down to the ideal papers for our research.

## Installation

You can install `petro.One` from Github with:

``` r
# install.packages("devtools")       # install *devtools* first

# install from the *master* release branch
devtools::install_github("f0nzie/petro.One")
```

Or, the latest development version with:

``` r
# install from the *develop* branch
devtools::install_github("f0nzie/petro.One", ref = "develop")
```

Or from [CRAN](https://cran.r-project.org/) with:

``` r
install.packages("petro.One")
```

The package dependencies can be installed with:

``` r
# install required packages
# from CRAN
install.packages(c("tm",
                   "wordcloud",
                   "urltools",
                   "RWeka",
                   "tidyverse",
                   "data.table",
                   "SnowBallC"
                   ))

# from BioConductor
source("https://bioconductor.org/biocLite.R")
biocLite("graph")
biocLite("Rgraphviz")
```

## What is behind the paper search

A typical OnePetro search URL would look like
this:

``` 
    https://www.onepetro.org/search?q=neural+network&peer_reviewed=&published_between=&from_year=&to_year=
    
```

You can see the search string for yourself after you hit enter on a
paper search, which could be explained like
this:

| Parameter           | Description                                                                                                                             |
| ------------------- | --------------------------------------------------------------------------------------------------------------------------------------- |
| domain              | <https://www.onepetro.org/>                                                                                                             |
| command             | search?                                                                                                                                 |
| q=                  | parameter that holds the query words. In the example above, it would be `q=neural+network`. As it is shown, it means search `any` word. |
| peer\_reviewed=     | parameter switch to get papers than have been only peer reviewed. When it has the value `on` means that is activated.                   |
| published\_between= | *parameter* switch that activates when `from_year` and `to_year` have numeric entries.                                                  |
| from\_year=         | *parameter* to enter the starting year of the search                                                                                    |
| to\_year=           | *parameter* to enter the end year of the search.                                                                                        |

There are additional parameters such
as:

| Parameter | Description                                                                                                                                                |
| --------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------- |
| start=    | indicate the starting page if the resulting search has several                                                                                             |
| rows=     | indicate the number of rows (papers) to display per page. In the web browser, the options are 10, 50 and 100. Off-browser it could be a number up to 1000. |
| sort=     | related to the selector `Sort By` with options `Relevance`, `Most recent` and `Highest rated`.                                                             |
| dc\_type= | indicates what type of document the paper is.                                                                                                              |

These are the type of documents:

    chapter
    conference-paper
    general
    journal-paper
    presentation
    media
    other
    standard

There are other parameters but they will not be used as often as the
ones already described.  
They key is build a search URL that is recognizable by OnePetro. To do
that I wrote a function `make_search_url` that does just that. Instead
of entering the search keywords, how will they be searched, year and
type of paper, we enter them from the R console.

# Examples

## Get the number of papers for the keyword *neural network*.

Let’s say we want to search the word **neural network**. How do we want
to perform that search? Paper that contain the word **neural** or the
word **network**, or papers that contain exactly the words **neural
network**? In the former, we would use the option `any`, in the latter,
`all`.

The option `how = "any"` means to search for papers that contain the
word `neural` or the word `network`. Let’s take a look at the difference
in returning results with `any` and `how = all` for the same keywords
`neural network`. Here we make of use of two functions in *petro.One*:
`make_search_url()` and `get_papers_count()`.

``` r
library(petro.One)
library(tidyverse)
#> -- Attaching packages -------------------------------------- tidyverse 1.2.1 --
#> v ggplot2 3.1.0     v purrr   0.2.5
#> v tibble  2.0.0     v dplyr   0.7.8
#> v tidyr   0.8.2     v stringr 1.3.1
#> v readr   1.3.1     v forcats 0.3.0
#> -- Conflicts ----------------------------------------- tidyverse_conflicts() --
#> x dplyr::filter() masks stats::filter()
#> x dplyr::lag()    masks stats::lag()

# search any word like "neural" or "network"
url_any <- make_search_url(query = "neural network", how = "any")
url_any
#> [1] "https://www.onepetro.org/search?q=neural+network&peer_reviewed=&published_between=&from_year=&to_year="
get_papers_count(url_any)
#> [1] 3931

# search for papers that have "neural" and "network" at the same time
url_all <- make_search_url(query = "neural network", how = "all")
url_all
#> [1] "https://www.onepetro.org/search?q=\"neural+network\"&peer_reviewed=&published_between=&from_year=&to_year="
get_papers_count(url_all)
#> [1] 3615
```

## Read papers from *from\_year* to *to\_year*

We can send a query where we specify the *starting year* and the *ending
year*. In this example the option `how = "all"` means to search papers
that contain **exactly** the words `neural network` as a difference to
`any` which means search for `any` occurrence of the words. Of course,
using `any` rather than `all` will yield many more results. We use two
petro.One functions: `make_search_url()` to build the OnePetro search
URL and `onepetro_page_to_dataframe()` to put the papers in a table.

``` r
library(petro.One)

# neural network papers from 1990 to 1999. Exact phrase
my_url <- make_search_url(query = "neural network", 
                          from_year = 1990, 
                          to_year   = 1999, 
                          how = "all")

df <- onepetro_page_to_dataframe(my_url)
df
#> # A tibble: 10 x 6
#>    book_title           paper_id  dc_type  authors              year source
#>    <fct>                <fct>     <fct>    <chr>               <int> <fct> 
#>  1 Deconvolution Using~ SEG-1996~ confere~ Essenreiter, Rober~  1996 SEG   
#>  2 Neural Network Stac~ SEG-1992~ confere~ Schmidt, Jumndyr, ~  1992 SEG   
#>  3 Neural Networks And~ SEG-1995~ confere~ Leggett, Miles, Br~  1995 SEG   
#>  4 First Break Picking~ SEG-1990~ confere~ Wagner, D.E., Amoc~  1990 SEG   
#>  5 Drill-Bit Diagnosis~ SPE-1955~ journal~ Arehart, R.A., Exx~  1990 SPE   
#>  6 Seismic Principal C~ SEG-1996~ confere~ Huang, Kou-Yuan, N~  1996 SEG   
#>  7 Reservoir Character~ SEG-1993~ confere~ An, P., University~  1993 SEG   
#>  8 Seismic Attribute C~ SEG-1993~ confere~ Johnston, David H.~  1993 SEG   
#>  9 Neural Networks For~ SEG-1993~ confere~ Hansen, Kim Vejlby~  1993 SEG   
#> 10 Neural Networks In ~ SEG-1991~ confere~ McCormack, Michael~  1991 SEG
```

And these are the terms that repeat more freqently:

``` r
term_frequency(df)
#> # A tibble: 27 x 2
#>    word       freq
#>    <chr>     <dbl>
#>  1 network      10
#>  2 neural       10
#>  3 use           5
#>  4 seismic       3
#>  5 pick          2
#>  6 analysi       1
#>  7 attribut      1
#>  8 break         1
#>  9 calibr        1
#> 10 character     1
#> # ... with 17 more rows
```

## Get papers by document type (*dc\_type*)

We can also get paper by the type of document. In OnePetro it is called
`dc_type`.

In this example we will be requesting only `conference-paper` type. Here
we add to `make_search_url()` the parameter `dc_type`. Note also that we
are adding another parameter `rows` to get 1000 rows instead of 10, 50
or 100 as the browser allows.

``` r
# specify document type = "conference-paper", rows = 1000

my_url <- make_search_url(query = "artificial intelligence", 
                          how = "all",
                          from_year = 1950, 
                          to_year   = 1980, 
                          dc_type = "conference-paper",
                          rows = 1000)

(pcount <- get_papers_count(my_url))
#> [1] 4
df <- onepetro_page_to_dataframe(my_url)
df
#> # A tibble: 3 x 6
#>   book_title              paper_id  dc_type   authors           year source
#>   <fct>                   <fct>     <fct>     <chr>            <int> <fct> 
#> 1 The Realities Of Radio~ OTC-1788~ conferen~ Moyer, James P.~  1973 OTC   
#> 2 Application of Digital~ SPE-2659~ conferen~ Hall, O.P. Jr.,~  1969 SPE   
#> 3 Computer Recognition O~ SPWLA-19~ conferen~ Schoonover, L.G~  1974 SPWLA
```

Only 4 on *artificial intelligence* from 1950 until 1980\!

## Finding the most frequent terms in *well test*

For this example we want to know about *conference papers* where the
words *well* and *test* are found together in the papers.

``` r
library(petro.One)

my_url <- make_search_url(query = "well test", 
                          dc_type = "conference-paper",
                          from_year = 2017, 
                          to_year   = 2018, 
                          how = "all")

get_papers_count(my_url)      # get the paper count
#> [1] 956
df <- read_multidoc(my_url)   # create a dataframe of papers
print(dim(df))                # dimension of the dataframe
#> [1] 956   6

(tf <- term_frequency(df))    # create a term frequency table
#> # A tibble: 1,974 x 2
#>    word       freq
#>    <chr>     <dbl>
#>  1 reservoir   325
#>  2 well        269
#>  3 field       223
#>  4 model       161
#>  5 fractur     160
#>  6 gas         152
#>  7 case        139
#>  8 use         136
#>  9 product     135
#> 10 studi       124
#> # ... with 1,964 more rows

# calculate the minimum frequency for, let's say, the first 20th term
min_freq <- min(head(tf, 20)$freq)
```

### Most frequent terms in *well test*

``` r
print(min_freq)
#> [1] 76
# plot most frequent terms
plot_bars(df, min.freq = min_freq)
```

![](man/figures/README-unnamed-chunk-8-1.png)<!-- -->

### How do the most frequent terms relate each other

Now, it is not enough for us to know what terms are the more repeating
but how those freqent terms relate to each other. In the following plot
you will see that the strength of the relationship between terms is
reflected by the thickness of the connection lines.

``` r
plot_relationships(df, min.freq = min_freq, threshold = 0.075)
```

![](man/figures/README-unnamed-chunk-9-1.png)<!-- -->

We can see that *wells* and *well* are connected strongly to
*horizontal*, *transient*, *pressure*, *flow*, *testing*, *reservoirs*,
*fracture*, and *analysis*. The rest of the words are frequent but not
very much connected.

## Summaries

The summary functions allow us to group the papers by a preferred group:

  - by type of document
  - by publisher
  - by publication
  - by year

This will give you a summary of the count not the papers themselves.

``` r
library(petro.One)

my_url <- make_search_url(query = "artificial intelligence", 
                          how = "all")
```

### By year of publication

| name       | value |
| :--------- | ----: |
| Since 2019 |     1 |
| Since 2018 |   416 |
| Since 2017 |   629 |
| Since 2016 |   807 |
| Since 2015 |   953 |
| Since 2014 |  1056 |
| Since 2013 |  1194 |
| Since 2012 |  1324 |
| Since 2011 |  1433 |
| Since 2010 |  1549 |


NEWS.md
## petro.One 0.2.3
* release to CRAN

## petro.One 0.2.2.9003
* remove Depends: R (>= 3.5.1) because crashes Travis

## petro.One 0.2.2.9002
* add Depends: R (>= 3.5.1)

## petro.One 0.2.2.9001
* remove devel from travis.yml
* add Depends: R (>= 3.4.0)

## petro.One 0.2.2.9000
* remove Depends. Causing errors in Travis
* adopting different numbering system for version

## petro.One 0.2.2
* add names to VignetteIndexEntry
* add Depends: R (>= 3.5.0)

## petro.One 0.2.1
* two notes from CRAN after submission
* 1 note from Debian, two notes from Windows
* change the travis.yml configuration file <- problem with devtools
* run in Mac osX, Windows and Mint 19

## petro.One 0.2.0
* new vignettes
* new function run_papers_search() that saves results to rda
* added new tests
* added new html offline files
* smaller README
* remove calls to rNodal.utils
* add examples in external functions
* add package SnowBallC
* add more functions to _pkgdown.yaml

## petro.One 0.1.3.9004
* bump version to 0.1.3.9004
* add plot of tri-gram
* enable man/ folder to git
* add parameter save_to_rda in function run_papers_search()

## petro.One 0.1.3.9003
* *20190109*
* document two new parameters for number of grams
* merge branch `Merge branch 'feature/arr-20180917-reservoir_ml_papers' into develop`.
* delete file arti.rda in project folder root.

## petro.One 0.1.3.9002
* merge *feature/result-item* with *develop*.
* vignettes disabled for the moment.

* *20180514*
* debug, fix problem with keyword "analytics". 
* add "verbose" parameter to read_multipage(). 
* add cat(sprintf()) for verbose to show page number being read and where it possible fails.
* additional logic for get_dc_type_raw() when data_itemid has 3 columns instead of 4. 
* new test script test_joining-analytics_error.R to confirm error is gone.

* *20180515*
* add skip() to **test_failing.R:25**
* updating test paper values
* document new parameter `verbose`.


## petro.One feature/result-item
* this branch out of `develop`.
* ignored branch: **feature/0.1.3.9001**
* copy draft files from GDrive: notebook, test file and script files to `./inst/result-item`
* *20180504*
* add script `install_packages.R` to document packages that need to be installed.
* *20180505*
* tested with R 3.4.4. passed with only one test enabled. moved from gdrive to github-oilgains: `test_multipage.R`. 
* add `mockery` to packages list.
* branch: **org-tests**
* only these tests doing `mock` tests with real data: 
*test_is_dctype_enabled-off.R*
*test_GTEsummary_by-off.R*
*test_onepetro_page_to_dataframe-off.R*
* mark tests that are offline
* 7:46 am - completing offline tests for read_multidoc().
* spli offline tests for read_multidoc() and read_multipage()
* Found discrepancies between numbers returned by the function and the values returned from OnePetro.
* *20180506*
* values from `paper_count()` do not match OnePetro. This is due to the addition of backslashes to the query when using the AND in the `how` parameter. To prevent this cat() the url generated by the function `make_search_url()`.
* add new test script `test_url.R`.
* make verbose=FALSE in join_keywords().
* rebuild read_multipage()
* add test for join_keywords()
* add test script for URL check. **always use cat()** of the url generated by R to remove the *backslashes*.
* fix problem with functions `get_papers_count` and `papers_by_*` when the url has the parameter `rows` in it. Added new funtion `check_unlimited_rows()` to address that.
* change variable name from "title_data" to "book_title". Build README.
* using *skip("skip")* in some tests.
*  Running 'testthat.R' [444s]
   * R CMD check results
   * 0 errors | 0 warnings | 0 notes


## petro.One 0.1.3.9000
* `20180227`
* fixing error when paper count is zero in function join_keywords().


## petro.One 0.1.3 20180218
* remove badge in README. Causing error.
* accepted by CRAN on 20180218 at 12:26 PM Singapore time.

## petro.One 0.1.2 20180217
* release to CRAN to fix error in tests
* returned by CRAN because a WARNING:

> * checking DESCRIPTION meta-information ... OK
* checking top-level files ... WARNING
Conversion of 'README.md' failed:
pandoc.exe: Could not fetch man/figures/README-unnamed-chunk-7-1.png
man/figures/README-unnamed-chunk-7-1.png: openBinaryFile: does not exist (No such file or directory)
* checking for left-over files ... OK

## petro.One 0.1.1.9000 
* `200817`
* add vignette for joining keywords
* add function `joining_keywords`
* add function to convert graphic foreign character to spaces in `get_term_document_matrix`.
* remove some tests to reduce test time.
* `20180211`
* notebook `water_injection_ml-9001.Rmd`advances the previous 9000, with a function that retrieves a dataframe of papers in addition to the paper count.
* `20180208`
* create function to join composite keywords (see water_injection_ml.Rmd)
* add check.attributes to function expect_equal_scale() 
* `20180207`
* add notebook for water injection, waterflood, machine learning, AI
* `20180124`
* use code in `gas_lift` notebook for article in LinkedIn
* add notebook that analyzes variation on the word gas lift.
* add some comments to Cantarell.
* `20180105`
* adding notebook Cantarell to find best paper for productin optimization.
* 20171102
* Link to "Thousands of Papers to Dataframe" not working.
* build site with pkgdown


## petro.One 0.1.1 20171102
* released to CRAN to fix two errors with the paper count
* using instead expect_equal to expect_gte in some unit tests that call for paper count.

## 0.1.0.9001 20171028-20171102
* bump to version 0.1.1 with patch
* change expect_equal to expect_gte because we have the expected values changing when OnePetro adds new papers.
* rename couple of test files to have the GTE words in them.
* fix issues with unit tests on papers. See cran-comments

## 0.1.0.9000 20171028
* remove commented lines from .travis.yml
* this dev version was incorrectly named 0.1.1.9000
* TODO:

R/url.R

#Loading the rvest package
library(rvest)

# Specifying the url for desired website to be scrapped
url <- "https://www.onepetro.org/search?start=0&q=neural+networks&from_year=&peer_reviewed=&published_between=&rows=999&to_year="

form_input <- list(dummy = "dummy", query = "?q=", peer_reviewed = "peer_reviewed=",
                   published_between = "published_between=",
                   from_year = "from_year=",
                   to_year = "to_year=",
                   start = "start=",
                   rows = "rows=",
                   dc_type = "dc_type=")

# Examples:
# https://www.onepetro.org/search?q=%22data+science%22&peer_reviewed=&published_between=&from_year=&to_year=
#




#' @title Read OnePetro web page given a query URL
#' @description Read a OnePetro webpage using a query URL. Uses xml2
#' function read_html
#' @param url char a query URL for OnePetro
#' @rdname read_onepetro
#' @importFrom xml2 read_html
#' @export
read_onepetro <- function(url) {
    #Reading the HTML code from the website
    read_html(url)
}


check_unlimited_rows <- function(url) {
    if (!is.na(urltools::param_get(url, "rows"))) {
        url <- urltools::param_remove(url, keys = c("rows"))
    }
    url
}

#' @title Number of paper for a given query
#' @description Obtains the number of papers being queried by the URL
#' @param url char a query URL for OnePetro
#' @rdname get_papers_count
#' @importFrom magrittr %>%
#' @importFrom rvest html_nodes html_text
#' @export
#' @examples
#' \dontrun{
#' # Example 1
#' url_1 <- make_search_url(query = "static gradient survey", how = "all")
#' get_papers_count(url_1)
#' #
#' # Example 2
#' url_2 <- make_search_url(query = "vertical lift performance", how = "all")
#' get_papers_count(url_2)
#' #
#' # Example 3
#' url_3 <- make_search_url(query = "inflow performance relationship", how = "all")
#' get_papers_count(url_3)
#' }
get_papers_count <- function(url) {
    url <- check_unlimited_rows(url)
    result <- xml2::read_html(url)

    papers <- result %>%
    html_nodes("h2") %>%
    html_text()

    # extract the numeric part of the results
    pattern <- "[\\d,]+(?= results.)"    # a number, including comma, before " results."
    m <- regexpr(pattern, papers[1], perl = TRUE)       # matched
    as.numeric(gsub(",", "", regmatches(papers[1], m))) # remove comma first
}


#' @title Make a search URL for OnePetro
#' @description Create a URL that works in OnePetro website
#' @param query char         any words that will be searched
#' @param start int          optional to set the starting paper
#' @param from_year int      optional to indicate starting year
#' @param peer_reviewed     logical optional, TRUE or FALSE
#' @param published_between logical automatic if from_year or to_year are on
#' @param rows int           optional. number of papers to retrieve. max=1000
#' @param to_year int        optional to indicate end year
#' @param dc_type char       optional to indicate if journal, conference paper
#' @param how char           default="any". "all" will match exact words
#' @export
#' @examples
#' \dontrun{
#' # Example 1
#' url_1 <- make_search_url(query = "flowing gradient survey", how = "all")
#' onepetro_page_to_dataframe(url_1)
#' # Example 2
#' url_2 <- make_search_url(query = "static  gradient survey", how = "all")
#' onepetro_page_to_dataframe(url_2)
#' # Example 3
#' url_3 <- make_search_url(query = "downhole flowrate measurement",
#'       how = "all", from_year = 1982, to_year = 2017)
#' onepetro_page_to_dataframe(url_3)
#' }
make_search_url <- function(query = NULL, start = NULL, from_year = NULL,
                            peer_reviewed = NULL,
                            published_between = NULL,
                            rows = NULL,
                            to_year = NULL,
                            dc_type = NULL,
                            how = "any") {

    website <- "https://www.onepetro.org"

    if (!is.null(start) || !is.null(rows)) {
        if (!is.null(rows) & is.null(start)) start = 0
        stopifnot(is.numeric(start), is.numeric(rows))
    }

    if (!is.null(from_year) && !is.null(to_year)) {
        stopifnot(is.numeric(from_year), is.numeric(to_year))
    }

    if (is.null(query)) {
        stop("search words not provided")
    } else {
        split_query <- unlist(strsplit(query, " "))
        if (length(split_query) > 1) {
            query <- paste(split_query, collapse = "+")
            query <- ifelse(how == "all", paste0("'", query, "'"), query)
            # print(query)
        }
    }

    if (!is.null(from_year) || !is.null(to_year)) {
        # use regex to validate year is between 1900 and 2099
        pattern <- "(?:(?:19|20)[0-9]{2})"
        if (!grepl(pattern, from_year, perl = TRUE) ||
            !grepl(pattern, to_year,   perl = TRUE)) stop("year not valid")
        # if valid year then turn on published_between
        published_between = "on"
        # if any of the *from* or *to* years are null replace with empty char
        if (is.null(from_year)) {
            from_year = ""
        }
        if (is.null(to_year)) {
            to_year = ""
        }
    }

    # peer_reviewed=on if TRUE; blank if unslected or FALSE
    if (is.null(peer_reviewed)) {
        peer_reviewed = ""
    } else {
        if (peer_reviewed) peer_reviewed = "on"
    }

    # document type
    if (!is.null(dc_type)) {
        valid_options <- c("conference-paper", "journal-paper",
                           "media", "general", "presentation", "chapter",
                           "other", "standard")
        # stop if it is not in the options
        if (!dc_type %in% valid_options) {
            msg <- sprintf("Option unknown. It must be one of [ %s ]",
                           paste(valid_options, collapse = ", "))
            stop(msg)
            # cat(valid_options, "\n")
        }
    }

    s_search  <- paste(website, "search", sep = "/")

    # these strings will need to join with the ampersand & at the tail
    s_query   <- paste0("?q=", query)
    s_peer    <- paste0("peer_reviewed=", peer_reviewed)
    s_publish <- paste0("published_between=", published_between)
    s_from    <- paste0("from_year=", from_year)
    s_to      <- paste0("to_year=", to_year)
    s_start   <- paste0("start=", start)
    s_rows    <- paste0("rows=", rows)
    s_type    <- paste0("dc_type=", dc_type)

    # url
    s_url <- list(websearch = s_search, query = s_query, peer = s_peer,
                  published_between = s_publish, from_year = s_from, to_year = s_to,
                  start = s_start, rows = s_rows, dc_type = s_type
    )

    for (i in 1:length(s_url)) {
        # cat(i, my_url[[i]], "\n")
        if (i == 1) joined <- s_url[[i]]
        if (i == 2) joined <- paste0(joined, s_url[[i]])
        if (i >=3 ) {
            if (s_url[[i]] == form_input[[i]] & i <= 6) {
                joined <- paste(joined, s_url[[i]], sep = "&")
            } else  if (s_url[[i]] != form_input[[i]]) {
                # cat(i, s_url[[i]], "\n")
                joined <- paste(joined, s_url[[i]], sep = "&")
            }
        }
    }

    q_url <- joined
    q_url <- gsub('"', "'", q_url)
    q_url <- gsub("'", '"', q_url)
    q_url
}








create_url <- function(start = NULL, query = NULL, from_year = NULL,
                       peer_reviewed = NULL,
                       published_between = NULL,
                       rows = NULL,
                       to_year = NULL,
                       how = "any") {

    website <- "https://www.onepetro.org"

    if (is.null(start)) {
        start = ""
    }
    if (is.null(query)) {
        stop("search words not provided")
    } else {
        split_query <- unlist(strsplit(query, " "))
        if (length(split_query) > 1) {
            query <- paste(split_query, collapse = "+")
            # use function shQuote to add extra quotes when we want how = "all"
            query <- ifelse(how == "all", shQuote(query), query)
            print(query)
        }
    }
    print(query)

    if (is.null(from_year)) {
        from_year = ""
    }
    if (is.null(peer_reviewed)) {
        peer_reviewed = ""
    }
    if (is.null(published_between)) {
        published_between = ""
    }
    if (is.null(rows)) {
        rows = ""
    } else {
        if(is.null(start)) start = 0
    }
    if (is.null(to_year)) {
        to_year = ""
    }

    s_search  <- paste(website, "search", sep = "/")
    s_q       <- paste0("?q=", query)
    s_peer    <- paste0("peer_reviewed=", peer_reviewed)
    s_publish <- paste0("published_between=", published_between)
    s_from    <- paste0("from_year=", from_year)
    s_to      <- paste0("to_year=", to_year)
    s_start   <- paste0("start=", start)
    s_rows    <- paste0("rows=", rows)

    url <- paste(s_q, s_peer, s_publish, s_from, s_to, sep = "&")
    url <- paste0(s_search, url)
    url
}


send_url <- function(url, how = "any") {
    #Reading the HTML code from the website
    read_html(url)
}
